### PR TITLE
Some extra developer conveniences for Docker image.

### PR DESCRIPTION
**Pull request by htuch**
* Refactor do_ci.sh to split the cmake and environment setup from the
  build and test. This makes it convenient for devs to invoke arbitrary
  make targets in the Docker image, e.g.

  docker run -t -i -v $PWD:/source lyft/envoy-build:latest /bin/bash -c \
    "cd /source && . ./ci/build_setup.sh && make fix_format"

* Add EXTRA_TEST_ARGS env var to pass addition flags to envoy-test. This
  is useful, for example, when applying a --gtest_filter to restrict the
  test run to only failing tests.

* Add a do_ci.sh fix_format target.

* Add gdb, strace to the image and a RUN_TEST_UNDER environment variable
  that can be used to have envoy-test run under these programs.

* UNIT_TEST_ONLY env var that will skip the non-envoy-test aspects of
  the test process.

Example command for debugging only tests matching *Dns*:

docker run -t -i -v $PWD:/source lyft/envoy-build:latest /bin/bash -c \
  "cd /source && UNIT_TEST_ONLY=1 RUN_TEST_UNDER='gdb --args' ./ci/do_ci.sh debug \
  --gtest_filter='*Dns*'

### CODE REVIEW

**Comment by mattklein123** [comment]
@htuch this is great. Can we document this somewhere? Would it be worthwhile to start a development markup page? (Not sure if it's worth putting this into the main sphinx docs but I could go either way).

**Comment by htuch** [comment]
I've added some MD documentation in ci/ for developers.

File: ci/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Developer use of CI Docker image
+
+The Docker image at `lyft/envoy-build:latest` is used for Travis CI checks. It
```
Comment (mattklein123) [inline ci/README.md]: small nit: as of https://github.com/lyft/envoy/pull/408 we are now pinning CI builds to a specific build image (easier to understand what is being tested, can upgrade an image and work on a PR, etc.). Do you mind just covering this briefly? (Basically that CI is now pinned, but latest will still work for dev if desired).

**Comment by rshriram** [comment]
I would also suggest linking to this doc from the sphinx docs (building and installation), and probably getting rid of that content. WDYT?

**Comment by rshriram** [comment]
So that there is one place to manage all dev setup documentation

File: ci/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+filesystem to be consumed and outputs build artifacts in `<SOURCE_DIR>/build`.
+The static Envoy binary can be found in `<SOURCE_DIR>/build/source/exe/envoy`.
+
+The `do_ci.sh` targets are:
+
+* `asan` &mdash; build and run tests with [AddressSanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer).
+* `coverage` &mdash; build and run tests, generating coverage information in `<SOURCE_DIR>/build/coverage.html`.
+* `debug` &mdash; build debug binary and run tests.
+* `docs` &mdash; build documentation, generated docs are found in `<SOURCE_DIR>/generated`.
+* `fix_format`&mdash; run `clang-format` on entire source tree.
```
Comment (rshriram) [inline ci/README.md]: Nit: clang format 3.6 specifically. I have gotten bit by this several times ;).

Comment (htuch) [inline ci/README.md]: Done.

File: ci/README.md (hunk 1 above)
Comment (htuch) [inline ci/README.md]: Done.

**Pull request merged**